### DISCUSSION TITLE
ESLint crash due to @graphql-eslint/graphql processor

### DISCUSSION THREAD

**Discussion by therealgilles** [eslint/eslint#16015] [Q&A]
The [@graphql-eslint/graphql](https://github.com/B2o5T/graphql-eslint) processor is making ESLint crash instead of reporting the syntax error. I am wondering what would make that happen. Any hint would be appreciated :)

```
ESLint: 8.17.0

SyntaxError: Missing semicolon. (19:20)
    at instantiate (.../node_modules/@babel/parser/lib/index.js:72:32)
    at constructor (.../node_modules/@babel/parser/lib/index.js:359:12)
    at Object.raise (.../node_modules/@babel/parser/lib/index.js:3342:19)
    at Object.semicolon (.../node_modules/@babel/parser/lib/index.js:4004:10)
    at Object.parseVarStatement (.../node_modules/@babel/parser/lib/index.js:15181:10)
    at Object.parseVarStatement (.../node_modules/@babel/parser/lib/index.js:10320:31)
    at Object.parseStatementContent (.../node_modules/@babel/parser/lib/index.js:14722:21)
    at Object.parseStatementContent (.../node_modules/@babel/parser/lib/index.js:10361:18)
    at Object.parseStatement (.../node_modules/@babel/parser/lib/index.js:14647:17)
    at Object.parseBlockOrModuleBlockBody (.../node_modules/@babel/parser/lib/index.js:15290:25)
```

**Comment by therealgilles** [comment]
Shouldn't ESLint not crash even if a plugin throws an error?

**Reply by mdjermanovic** [reply]
> Shouldn't ESLint not crash even if a plugin throws an error?

Processors and rules are not expected to throw errors. When a processor or rule throws an error that means that something unpredicted has happened, and thus crashing is the intended behavior.

Parsers are expected to throw errors on invalid syntax. ESLint [catches](https://github.com/eslint/eslint/blob/c1d02400ec67c0a3de0fc77f750ad08542e96034/lib/linter/linter.js#L835-L852) those errors and treats them as lint problems.

> ESLint: 8.17.0
> 
> SyntaxError: Missing semicolon. (19:20)
>     at instantiate (.../node_modules/@babel/parser/lib/index.js:72:32)
>     at constructor (.../node_modules/@babel/parser/lib/index.js:359:12)
>     at Object.raise (.../node_modules/@babel/parser/lib/index.js:3342:19)
>     at Object.semicolon (.../node_modules/@babel/parser/lib/index.js:4004:10)
>     at Object.parseVarStatement (.../node_modules/@babel/parser/lib/index.js:15181:10)
>     at Object.parseVarStatement (.../node_modules/@babel/parser/lib/index.js:10320:31)
>     at Object.parseStatementContent (.../node_modules/@babel/parser/lib/index.js:14722:21)
>     at Object.parseStatementContent (.../node_modules/@babel/parser/lib/index.js:10361:18)
>     at Object.parseStatement (.../node_modules/@babel/parser/lib/index.js:14647:17)
>     at Object.parseBlockOrModuleBlockBody (.../node_modules/@babel/parser/lib/index.js:15290:25)

Per the stack trace, this does look like a parsing error. but it isn't clear if the parser was called directly from ESLint or maybe from a processor. A repro repo would be helpful to figure out why this results in crashing instead of reporting the parsing error as a lint problem. Debug info ([`--debug`](https://eslint.org/docs/user-guide/command-line-interface#--debug) flag) might also be helpful.

**Reply by therealgilles** [reply]
Thank you for the prompt reply :)

Sorry for not adding more details in my question. What happens is @graphql-eslint/graphql is specified as processor to extract the GraphQL strings from the code. It, in turn, calls @graphql-tools/graphql-tag-pluck, which itself calls @babel/parser, which throws the error. I don't know much about parsers vs processors at this point. Is this not a valid flow? how can a processor be expected not to ever throw an error?

**Reply by mdjermanovic** [reply]
> how can a processor be expected not to ever throw an error?

That's a good question. I believe that the reason why processor errors that would be handled similarly to parser errors are not supported is that, when used with a processor, ESLint is responsible for validating code blocks from the file, but not the file as a whole. When a file is syntactically invalid so that the processor cannot parse it, the processor should return an empty array (which means no code blocks were found) because there's nothing for ESLint to work on.

@btmills Is this reasoning correct?

**Reply by mdjermanovic** [reply]
> What happens is @graphql-eslint/graphql is specified as processor to extract the GraphQL strings from the code. It, in turn, calls @graphql-tools/graphql-tag-pluck, which itself calls @babel/parser, which throws the error.

[`preprocess`](https://github.com/B2o5T/graphql-eslint/blob/29b52b0d1b6778c50572f34e098a6a726a5201f0/packages/plugin/src/processor.ts#L14) returns extracted `.graphql` code blocks, and also the entire file content (`code`) as another code block. I _think_ it should catch errors from `@graphql-tools/graphql-tag-pluck`, and in that case return only the code block that represents the entire file (`code`), which would then be linted and cause a parsing error that doesn't break the whole linting.

**Reply by therealgilles** [reply]
I find it odd that it would be okay for ESLint to crash, also making it difficult to figure out why it crashed. In my mind, I would think you would want ESLint to never crash and in the present case, report that the @graphql-eslint/graphql preprocessor threw an error, so that as a user I understand the issue is with the preprocessor and I can address it with its developers. You could also potentially have ESLint keep going ignoring the faulty preprocessor. It would make for a better UX, make ESLint more robust, and lighten reports similar to this one.

**Reply by btmills** [reply]
> ESLint is responsible for validating code blocks from the file, but not the file as a whole.

@mdjermanovic this is true currently, though it's expected to change as part of https://github.com/eslint/eslint/issues/14745. Historically, I think we assumed that processors would be fairly liberal in what they accept, skipping over most content without parsing and selecting only the target blocks. The GraphQL processor appears to be strictly parsing the entire file with Babel and getting caught there.

I agree it would be better to report syntax errors in processed files just like we report JS syntax errors as lint messages It would be fairly simple to wrap the [`preprocess()` call](https://github.com/eslint/eslint/blob/ae4b449719d496fd611903d596341ec4c1d81c16/lib/linter/linter.js#L1793) in a `try` so that the `catch` can return a `LintMessage` showing the file-level syntax error much like [`parse()` does](https://github.com/eslint/eslint/blob/ae4b449719d496fd611903d596341ec4c1d81c16/lib/linter/linter.js#L844-L851).

**Reply by JounQin** [reply]
Is this going to be implemented? I'd like to take it if so.

**Reply by btmills** [reply]
I just opened https://github.com/eslint/eslint/issues/16104 summarizing the issue and my opinion on the solution. Assuming nobody objects in the next few days, feel free to tackle it!

**Comment by therealgilles** [comment]
I made a bit of progress tracking down this issue. ESLint still crashes on the command line but only crashes inside VSCode if I have the following in .vscode/settings.json:

```JS 
  // I have to comment this out, otherwise ESLint crashes inside VSCode,
  // displaying popup windows one after the other on each crash.
  "eslint.validate": [
    "javascript",
    "javascriptreact",
    "typescript",
    "typescriptreact",
    "graphql"
  ],
```
